Microsoft.Toolkit.Uwp.UI.Controls/WrapPanel/WrapPanel.cs

```diff
@@ -0,0 +1,157 @@
+    /// </summary>
+    public partial class WrapPanel : Panel
+    {
+        private List<double> _vSize;
+
+        /// <summary>
+        /// Gets or sets the orientation of the WrapPanel, Horizontal or vertical means that child controls will be added horizontally until the width of the panel can't fit more control then a new row is added to fit new horizontal added child controls, vertical means that child will be added vertically until the height of the panel is recieved then a new column is added
+        /// </summary>
+        public Orientation Orientation
+        {
+            get { return (Orientation)GetValue(OrientationProperty); }
+            set { SetValue(OrientationProperty, value); }
+        }
+
+        /// <summary>
+        /// Identifies the <see cref="Orientation"/> dependency property.
+        /// </summary>
+        public static readonly DependencyProperty OrientationProperty =
+            DependencyProperty.Register(
+                "Orientation",
+                typeof(Orientation),
+                typeof(WrapPanel),
+                new PropertyMetadata(Orientation.Horizontal));
+
+        /// <inheritdoc />
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            _vSize = new List<double> { 0.0 };
+            var totalMeasure = new UvMeasure();
+            var parentMeasure = new UvMeasure(Orientation, availableSize.Width, availableSize.Height);
+            var lineMeasure = new UvMeasure();
+            foreach (var child in Children)
+            {
+                child.Measure(availableSize);
+
+                var currentMeasure = new UvMeasure(Orientation, child.DesiredSize.Width, child.DesiredSize.Height);
+
+                if (parentMeasure.U > currentMeasure.U + lineMeasure.U)
+                {
+                    lineMeasure.U += currentMeasure.U;
+                    lineMeasure.V = Math.Max(lineMeasure.V, currentMeasure.V);
+                }
+                else
+                {
+                    // new line should be added
+                    // to get the max U to provide it correctly to ui width ex: ---| or -----|
+                    totalMeasure.U = Math.Max(lineMeasure.U, totalMeasure.U);
+                    totalMeasure.V += lineMeasure.V;
+                    _vSize.Add(lineMeasure.V);
+
+                    // if the next new row still can handle more controls
+                    if (parentMeasure.U > currentMeasure.U)
+                    {
+                        // set lineMeasure initial values to the currentMeasure to be calculated later on the new loop
+                        lineMeasure = currentMeasure;
+                    }
+
+                    // the control will take one row alone
+                    else
```

Comment:
they are totally 2 different cases.
the first if the new add line can handle more control along with the current one
the second if the new line will have only one control so add the control U then add a new line.